Userland/xargs.cpp

```diff
@@ -0,0 +1,285 @@
+            perror("fopen");
+            return 1;
+        }
+        is_stdin = false;
+    }
+
+    StringBuilder builder;
+    Vector<char*> child_argv;
+
+    int devnull_fd = 0;
+
+    if (is_stdin) {
+        devnull_fd = open("/dev/null", O_RDONLY | O_CLOEXEC);
+        if (devnull_fd < 0) {
+            perror("open");
+            return 1;
+        }
+    }
+
+    size_t total_command_length = 0;
+    size_t lines_used_for_this_command = 0;
+
+    auto fail = read_lines(fp, entry_separator, [&](StringView line) {
+        if (line.ends_with('\n'))
+            line = line.substring_view(0, line.length() - 1);
+
+        if (line.is_empty())
+            return Continue;
+
+        if (lines_used_for_this_command == 0) {
+            child_argv.ensure_capacity(argument_parts.size());
+            for (auto& parts : argument_parts) {
+                builder.clear();
+                builder.join(line, parts);
+                auto string = builder.build();
+                total_command_length += string.length();
+                child_argv.unchecked_append(strdup(string.characters()));
+            }
+            ++lines_used_for_this_command;
+        } else {
+            if ((max_lines > 0 && lines_used_for_this_command > max_lines - 1) || total_command_length + line.length() + 1 >= max_bytes) {
+                if (!run_command(child_argv, verbose, is_stdin, devnull_fd))
+                    return Stop;
+                lines_used_for_this_command = 0;
+                total_command_length = 0;
+                return Unget;
+            } else {
+                child_argv.append(strndup(line.characters_without_null_termination(), line.length()));
+                ++lines_used_for_this_command;
+            }
+        }
+
+        return Continue;
+    });
+
+    if (!child_argv.is_empty())
+        run_command(child_argv, verbose, is_stdin, devnull_fd);
+
+    if (!is_stdin)
+        fclose(fp);
+
+    return fail ? 1 : 0;
+}
+
+bool read_lines(FILE* fp, char entry_separator, Function<Decision(StringView)> callback)
+{
+    char* line = nullptr;
+    size_t buffer_size = 0;
+
+    bool fail = false;
+    char* previous_line = nullptr;
+    size_t previous_line_size = 0;
+
+    for (;;) {
+        line = nullptr;
+        buffer_size = 0;
+
+        ssize_t line_size;
+
+        if (previous_line) {
+            line_size = previous_line_size;
+            line = previous_line;
+            previous_line = nullptr;
+        } else {
+            line_size = getdelim(&line, &buffer_size, entry_separator, fp);
+        }
+
+        if (line_size < 0) {
+            // getdelim() will return -1 and set errno to 0 on EOF.
+            if (errno != 0) {
+                perror("getdelim");
+                fail = true;
+            }
+            break;
+        }
+
+        previous_line = line;
+        previous_line_size = line_size;
+
+        auto decision = callback({ line, (size_t)line_size });
+        free(line);
```

Comment:
You can't `free()` it yet, `previous_line` refers to it.

An idea how to simplify this further:

```c++
while (true) {
    char* line;
    getdelim(...);

    Decision decision;
    do {
        decision = callback(line);
        if (decision == Stop) {
            free(line);
            return;
        }
    } while (decision == Unget);

    free(line);
}
```

and no need for separate `previous_*` then.

Reply:
I like that, sure, let's go with that.